pkg/buildkit/worker/leasemanager.go

```diff
@@ -0,0 +1,345 @@
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	pod, err := m.nextAvailablePod(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	addr, err := m.buildURL(ctx, pod.Name)
+	if err != nil {
+		return "", err
+	}
+
+	pac, err := corev1ac.ExtractPod(pod, fieldManagerName)
+	if err != nil {
+		return "", fmt.Errorf("cannot extract pod config: %w", err)
+	}
+
+	pac.WithAnnotations(map[string]string{
+		leasedAnnotation:    "true",
+		managerIDAnnotation: m.uuid,
+	})
+	delete(pac.Annotations, expiryTimeAnnotation)
+
+	m.log.Info("Applying pod metadata changes", "annotations", pac.Annotations)
+	if _, err = m.podClient.Apply(ctx, pac, metav1.ApplyOptions{FieldManager: fieldManagerName}); err != nil {
+		return "", fmt.Errorf("cannot update pod metadata: %w", err)
+	}
+
+	return addr, nil
+}
+
+func (m *podLeaseManager) Release(ctx context.Context, addr string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.log.Info("Parsing lease addr", "addr", addr)
+	u, err := url.ParseRequestURI(addr)
+	if err != nil {
+		return fmt.Errorf("failed to parse lease addr: %w", err)
+	}
+	podName := strings.Split(u.Host, ".")[0]
+
+	m.log.Info("Querying for pod", "name", podName, "namespace", m.namespace)
+	pod, err := m.podClient.Get(ctx, podName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			err = fmt.Errorf("addr %q is not allocated: %w", addr, err)
+		}
+
+		return err
+	}
+
+	pac, err := corev1ac.ExtractPod(pod, fieldManagerName)
+	if err != nil {
+		return fmt.Errorf("cannot extract pod config: %w", err)
+	}
+
+	pac.WithAnnotations(map[string]string{
+		expiryTimeAnnotation: time.Now().Add(m.podMaxIdleTime).Format(time.RFC3339),
+	})
+	delete(pac.Annotations, leasedAnnotation)
+	delete(pac.Annotations, managerIDAnnotation)
+
+	m.log.Info("Applying pod metadata changes", "annotations", pac.Annotations)
+	if pod, err = m.podClient.Apply(ctx, pac, metav1.ApplyOptions{FieldManager: fieldManagerName}); err != nil {
+		return fmt.Errorf("cannot update pod metadata: %w", err)
+	}
+
+	go func() {
+		m.ptch <- podTransport{pod: pod, err: nil}
+	}()
+
+	return nil
+}
+
+func (m *podLeaseManager) nextAvailablePod(ctx context.Context) (*corev1.Pod, error) {
+	select {
+	case pt := <-m.ptch:
+		return pt.pod, pt.err
+	default:
+		go func(ctx context.Context) {
+			pod, err := m.getUnleasedOrScale(ctx)
+			m.ptch <- podTransport{pod: pod, err: err}
+		}(ctx)
+	}
+
+	res := <-m.ptch
+	return res.pod, res.err
+}
+
+func (m *podLeaseManager) getUnleasedOrScale(ctx context.Context) (*corev1.Pod, error) {
+	m.log.Info("Checking for unleased pods")
+	pod, err := m.getUnleasedPod(ctx)
+	if err == nil {
+		return pod, err
+	}
+
+	if !errors.Is(err, errNoUnleasedPods) {
+		return nil, fmt.Errorf("unexpected leasing error: %w", err)
+	}
+
+	m.log.Info("Scaling cluster, no unleased pods found")
+	if err = m.scaler.Scale(ctx, 1); err != nil {
+		return nil, err
+	}
+
+	m.log.Info("No unleased pods found, scaling workload")
```

Comment:
I'm not sure I understand this log message in this context

Reply:
whoops, good catch. i meant to remove this.